pylops/waveeqprocessing/_kirchhoff_cuda.py

```diff
@@ -0,0 +1,519 @@
+        rix_d = cuda.to_device(rix)
+        self.const_inputs = (ns_d, nr_d, nt_d, ni_d, nz_d, dt_d, aperturemin_d, aperturemax_d, angleaperturemin_d,
+                             angleaperturemax_d, vel_d, aperturetap_d, six_d, rix_d)
+
+        self.trav_recs_d_global = cuda.to_device(trav_recs)
+        self.angles_recs_d_global = cuda.to_device(angle_recs)
+        self.trav_srcs_d_global = cuda.to_device(trav_srcs)
+        self.angles_srcs_d_global = cuda.to_device(angle_srcs)
+        self.amp_srcs_d_global = cuda.to_device(amp_srcs)
+        self.amp_recs_d_global = cuda.to_device(amp_recs)
+
+    @staticmethod
+    @cuda.jit
+    def _trav_kirch_matvec_cuda(x, y, nsnr, nt, ni, itrav, travd):
+        isrcrec = cuda.grid(1)
+        if nsnr > isrcrec:
+            for ii in range(ni):
+                itravisrcrec = itrav[:, isrcrec]
+                travdisrcrec = travd[:, isrcrec]
+                itravii = itravisrcrec[ii]
+                travdii = travdisrcrec[ii]
+                if 0 <= itravii < nt - 1:
+                    ind1 = (isrcrec, itravii)
+                    val1 = x[ii] * (1 - travdii)
+                    ind2 = (isrcrec, itravii + 1)
+                    val2 = x[ii] * travdii
+                    cuda.atomic.add(y, ind1, val1)
+                    cuda.atomic.add(y, ind2, val2)
+
+    @staticmethod
+    @cuda.jit
+    def _trav_kirch_rmatvec_cuda(x, y, nsnr, nt, ni, itrav, travd):
+        isrcrec = cuda.grid(1)
+        if nsnr > isrcrec:
+            for ii in range(ni):
+                itravii = itrav[ii]
+                travdii = travd[ii]
+                itravisrcrecii = itravii[isrcrec]
+                travdisrcrecii = travdii[isrcrec]
+                if 0 <= itravisrcrecii < nt - 1:
+                    vii = (
+                            x[isrcrec, itravisrcrecii] * (1 - travdisrcrecii)
+                            + x[isrcrec, itravisrcrecii + 1] * travdisrcrecii
+                    )
+                    cuda.atomic.add(y, ii, vii)
+
+    @staticmethod
+    @cuda.jit
+    def _travsrcrec_kirch_matvec_cuda(x, y, ns, nr, nt, ni, dt, trav_srcs, trav_recs):
+        isrc, irec = cuda.grid(2)
+        if ns > isrc and nr > irec:
+            for ii in range(ni):
+                travisrc = trav_srcs[ii, isrc]
+                travirec = trav_recs[ii, irec]
+                trav = travisrc + travirec
+                itravii = int(trav / dt)
+                travdii = trav / dt - itravii
+                if 0 <= itravii < nt - 1:
+                    ind1 = (isrc * nr + irec, itravii)
+                    val1 = x[ii] * (1 - travdii)
+                    ind2 = (isrc * nr + irec, itravii + 1)
+                    val2 = x[ii] * travdii
+                    cuda.atomic.add(y, ind1, val1)
+                    cuda.atomic.add(y, ind2, val2)
+
+    @staticmethod
+    @cuda.jit
+    def _travsrcrec_kirch_rmatvec_cuda(x, y, ns, nr, nt, ni, dt, trav_srcs, trav_recs):
+        irec, isrc = cuda.grid(2)
```

Comment:
why this is swapped compared to line 136?

Reply:
Thank you for noticing! The tests have passed for a uniform number of sources and receivers.

However, I anticipate that if we vary these numbers, the kernel launch and setup may not adequately cover both. So, this line should remain consistent across all 2D grid launch calls and setups. I’ll make sure to enhance it accordingly.

Reply:
Ok, great! A good reccomandation is to never test codes with forced `symmetries` as many bugs could pass unnoticed (i sometimes make this mistake myself 😆 )